internal/streamingnode/server/service/handler/consumer/consume_server.go

```diff
@@ -0,0 +1,189 @@
+		closeCh:       make(chan struct{}),
+	}, nil
+}
+
+// ConsumeServer is a ConsumeServer of log messages.
+type ConsumeServer struct {
+	scanner       wal.Scanner
+	consumeServer *consumeGrpcServerHelper
+	logger        *log.MLogger
+	closeCh       chan struct{}
+}
+
+// Execute executes the consumer.
+func (c *ConsumeServer) Execute() error {
+	// recv loop will be blocked until the stream is closed.
+	// 1. close by client.
+	// 2. close by server context cancel by return of outside Execute.
+	go c.recvLoop()
+
+	// Start a send loop on current goroutine.
+	// the loop will be blocked until:
+	// 1. the stream is broken.
+	// 2. recv arm recv close signal.
+	// 3. scanner is quit with expected error.
+	return c.sendLoop()
+}
+
+// sendLoop sends the message to client.
+func (c *ConsumeServer) sendLoop() (err error) {
+	defer func() {
+		if err := c.scanner.Close(); err != nil {
+			c.logger.Warn("close scanner failed", zap.Error(err))
+		}
+		if err != nil {
+			c.logger.Warn("send arm of stream closed by unexpected error", zap.Error(err))
+			return
+		}
+		c.logger.Info("send arm of stream closed")
+	}()
+	// Read ahead buffer is implemented by scanner.
+	// Do not add buffer here.
+	for {
+		select {
+		case msg, ok := <-c.scanner.Chan():
+			if !ok {
+				return status.NewInner("scanner error: %s", c.scanner.Error())
+			}
+			// Send Consumed message to client and do metrics.
+			messageSize := msg.EstimateSize()
+			if err := c.consumeServer.SendConsumeMessage(&streamingpb.ConsumeMessageReponse{
+				Id: &streamingpb.MessageID{
+					Id: msg.MessageID().Marshal(),
+				},
+				Message: &streamingpb.Message{
+					Payload:    msg.Payload(),
+					Properties: msg.Properties().ToRawMap(),
+				},
+			}); err != nil {
+				return status.NewInner("send consume message failed: %s", err.Error())
+			}
+			metrics.StreamingNodeConsumeBytes.WithLabelValues(
+				paramtable.GetStringNodeID(),
+				c.scanner.Channel().Name,
+				strconv.FormatInt(c.scanner.Channel().Term, 10),
+			).Observe(float64(messageSize))
+		case <-c.closeCh:
+			c.logger.Info("close channel notified")
+			if err := c.consumeServer.SendClosed(); err != nil {
+				c.logger.Warn("send close failed", zap.Error(err))
+				return status.NewInner("close send server failed: %s", err.Error())
+			}
+			return nil
+		case <-c.consumeServer.Context().Done():
+			return c.consumeServer.Context().Err()
+		}
+	}
+}
+
+// recvLoop receives messages from client.
+func (c *ConsumeServer) recvLoop() (err error) {
+	defer func() {
+		if err != nil {
+			c.logger.Warn("recv arm of stream closed by unexpected error", zap.Error(err))
+			return
+		}
+		c.logger.Info("recv arm of stream closed")
+	}()
+
+	for {
+		req, err := c.consumeServer.Recv()
+		if err == io.EOF {
+			return nil
+		}
+		if err != nil {
+			return err
+		}
+		switch req := req.Request.(type) {
+		case *streamingpb.ConsumeRequest_Close:
+			close(c.closeCh)
```

Comment:
Wait recv io.EOF or error is best to way to avoid server resource leak.
I will move the closeCh into defer function, and just do nothing when recv Close operation.